### PR TITLE
Autocomplete for child and parent relationships on work edit form.

### LINKED ISSUE #1563: UI Create _find_work_widget partial
### Wireframes:
- [in context of New Work](https://wiki.duraspace.org/display/hydra/New+Work+-+tab+Relationships+-+Phase+1+-+UI+Wireframes#mockup_p1-work-new-tab_relationships) - Relationship Tab - sections This Work in Other Works, Other Works in This Work, Related Works
- [in context of Edit Work](https://wiki.duraspace.org/display/hydra/Edit+Work+-+Relationships+Tab+-+Phase+1+-+UI+Wireframes) - Relationship Tab - sections This Work in Other Works, Other Works in This Work, Related Works
- [in context of Batch Upload](https://wiki.duraspace.org/display/hydra/Batch+New+Works+-+tab+Relationships+-+Phase+1+-+UI+Wireframes) - Relationship Tab - sections These New Works in Other Works, Other Works in These New Works, Related Works
### Usage:

Used to find and select an existing work.  Options for finding a work:
- user begins typing the name of the work and auto-complete suggests works
- user chooses to search for works and can browse a list of all works with pagination
### Implementation:
- File: /app/views/generic_works/_find_work_widget.html.erb
- Create a partial for a view that includes the find_work_widget.
  - need to be able to name the field associated with the widget
  - need to be able to return a value for the field based on the user's selection
- Add class/methods that support auto-complete
- Add class/methods that support browse and select

### PR DESCRIPTION
**Pull request by andjsmit**
Fixes #1563

Creates autocomplete fields for child and parent relationship fields on work edit form using Question Your Authority and a new search builder.

@projecthydra/sufia-code-reviewers

### CODE REVIEW

File: app/models/qa/authorities/local/find_works.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module Qa::Authorities
+  class Local::FindWorks < ActiveFedora::Base
```
Comment (jcoyne) [inline app/models/qa/authorities/local/find_works.rb]: Why does this extend ActiveFedora::Base?  Shouldn't it extend `Qa::Authorities::Base`?

File: app/models/qa/authorities/local/find_works.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      @q = q
+    end
+
+    def search(_q, controller)
+      repo = CatalogController.new.repository
+      builder = Sufia::FindWorksSearchBuilder.new(controller)
+      response = repo.search(builder)
+      docs = response.documents
+      results = []
+      docs.each do |doc|
```
Comment (jcoyne) [inline app/models/qa/authorities/local/find_works.rb]: Instead of doing `each` could you use `map` instead? Then you can also remove line 14 and 20

File: app/models/qa/authorities/local/find_works.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module Qa::Authorities
+  class Local::FindWorks < ActiveFedora::Base
+    attr_reader :q
+
+    def initialize(q)
+      @q = q
```
Comment (jcoyne) [inline app/models/qa/authorities/local/find_works.rb]: Why are you assigning `@q`, or even defining `initialize`? This should be handled by the base.

File: app/search_builders/sufia/find_works_search_builder.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  include CurationConcerns::FilterByType
+
+  def initialize(context)
+    super(context)
+    @id = context.params[:id]
+    @q = context.params[:q]
+  end
+
+  self.default_processor_chain += [:filter_on_title, :show_only_resources_deposited_by_current_user]
+  self.default_processor_chain += [:show_only_other_works, :show_only_works_not_child, :show_only_works_not_parent]
```
Comment (jcoyne) [inline app/search_builders/sufia/find_works_search_builder.rb]: please combine lines 11 & 12 and move them above the initializer.

File: app/search_builders/sufia/find_works_search_builder.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
+class Sufia::FindWorksSearchBuilder < Sufia::SearchBuilder
+  include Sufia::MySearchBuilderBehavior
```
Comment (jcoyne) [inline app/search_builders/sufia/find_works_search_builder.rb]: Add a comment here about why you're including this.  Something like `# Only search for works that we have edit access to`

File: app/models/qa/authorities/local/find_works.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+module Qa::Authorities
```
Comment (jcoyne) [inline app/models/qa/authorities/local/find_works.rb]: I don't think this belongs in `app/models`, how about `app/authorities` instead? I also don't think we want it in the `Qa` namespace, how about `Sufia`?

File: spec/models/qa/authorities/local/find_works_spec.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  subject { service.search(q, controller) }
+
+  describe '#search' do
+    let!(:work1) { create(:generic_work, :public, title: ['foo'], user: user1) }
+    let!(:work2) { create(:generic_work, :public, title: ['foo foo'], user: user1) }
+    let!(:work3) { create(:generic_work, :public, title: ['foo foo foo'], user: user2) }
+    let!(:work4) { create(:generic_work, :public, title: ['bar'], user: user1) }
+
+    it 'shows a list of other works deposited by current user' do
```
Comment (jcoyne) [inline spec/models/qa/authorities/local/find_works_spec.rb]: Can you add a test that shows parent works/child works are properly excluded?

File: app/models/qa/authorities/local/find_works.rb (hunk 6 above)
Comment (andjsmit) [inline app/models/qa/authorities/local/find_works.rb]: Would `app/authorities/sufia/local/find_works.rb` be correct ?

Comment (jcoyne) [inline app/models/qa/authorities/local/find_works.rb]: I think that would be fine.

File: app/models/qa/authorities/local/find_works.rb (hunk 3 above)
Comment (andjsmit) [inline app/models/qa/authorities/local/find_works.rb]: Without this init, I get an error : `ArgumentError (wrong number of arguments (given 1, expected 0))` even after extending `Qa::Authorities::Base` as requested above.

File: app/models/qa/authorities/local/find_works.rb (hunk 6 above)
Comment (jcoyne) [inline app/models/qa/authorities/local/find_works.rb]: Or  `app/authorities/sufia/authorities/local/find_works.rb` if you care going to put it in the `Sufia::Authoritities::Local` namespace.

File: app/search_builders/sufia/find_works_search_builder.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    solr_parameters[:fq] += [
+      "-" + ActiveFedora::SolrQueryBuilder.construct_query_for_ids([@id])
+    ]
+  end
+
+  def show_only_works_not_child(solr_parameters)
+    ids = ActiveFedora::SolrService.query("{!field f=id}#{@id}", fl: "member_ids_ssim").flat_map { |x| x.fetch("member_ids_ssim", []) }
+    solr_parameters[:fq] ||= []
+    solr_parameters[:fq]  += [
+      "-" + ActiveFedora::SolrQueryBuilder.construct_query_for_ids([ids])
```
Comment (jeremyf) [inline app/search_builders/sufia/find_works_search_builder.rb]: If I'm reading this correctly, the `ids` from line 27 should be an array. Then you are wrapping the `ids` in an array.

I believe you may want `ActiveFedora::SolrQueryBuilder.construct_query_for_ids(Array.wrap(ids))`

Comment (jcoyne) [inline app/search_builders/sufia/find_works_search_builder.rb]: I think it's safe to say ids is an array (output of `#map`), there's no need to wrap it.

File: app/search_builders/sufia/find_works_search_builder.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+  def show_only_other_works(solr_parameters)
+    solr_parameters[:fq] ||= []
+    solr_parameters[:fq] += [
+      "-" + ActiveFedora::SolrQueryBuilder.construct_query_for_ids([@id])
+    ]
+  end
+
+  def show_only_works_not_child(solr_parameters)
+    ids = ActiveFedora::SolrService.query("{!field f=id}#{@id}", fl: "member_ids_ssim").flat_map { |x| x.fetch("member_ids_ssim", []) }
```
Comment (jcoyne) [inline app/search_builders/sufia/find_works_search_builder.rb]: make sure you pass `:rows` to `ActiveFedora::SolrService.query` or it'll only return the first 10.

Comment (andjsmit) [inline app/search_builders/sufia/find_works_search_builder.rb]: I set `rows: 0` . Is this the correct ay to get all results?

Comment (andjsmit) [inline app/search_builders/sufia/find_works_search_builder.rb]: @jcoyne Never mind, that breaks it. So, just set it to a large number? `rows: 10000` ?

Comment (jcoyne) [inline app/search_builders/sufia/find_works_search_builder.rb]: @andjsmit yep. 10000 ought to be good.

File: app/authorities/sufia/authorities/local/find_works.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
+module Sufia::Authorities
+  class Local::FindWorks < Qa::Authorities::Base
+    attr_reader :q
+
+    def initialize(q)
```
Comment (jeremyf) [inline app/authorities/sufia/authorities/local/find_works.rb]: What is `q`. The terse variable name is confusing and not obvious.

Comment (andjsmit) [inline app/authorities/sufia/authorities/local/find_works.rb]: I assumed it was the required query variable for the QA.  @jcoyne had mentioned removing it. But in doing so, the code crashes.

Comment (jeremyf) [inline app/authorities/sufia/authorities/local/find_works.rb]: @andjsmit But what does the `q` variable mean?

Comment (andjsmit) [inline app/authorities/sufia/authorities/local/find_works.rb]: Ok. I did a little digging. It is not the actual query as I thought. It appears to be the type? When I check the value of "q" in the init, I get `works`. So should I be calling it "type"?

File: app/authorities/sufia/authorities/local/find_works.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
+module Sufia::Authorities
+  class Local::FindWorks < Qa::Authorities::Base
+    attr_reader :q
```
Comment (jcoyne) [inline app/authorities/sufia/authorities/local/find_works.rb]: I don't see this attribute ever accessed. Is it needed?

Comment (andjsmit) [inline app/authorities/sufia/authorities/local/find_works.rb]: @jcoyne Without it, a wrong number of arguments error occurs (given 1, expected 0)

```
 def instance
    klass.new(@subauthority)
```
in questioning_authority/lib/qa/authorities/local/registry.rb.

Comment (jcoyne) [inline app/authorities/sufia/authorities/local/find_works.rb]: You are talking about the parameter to initialize.  I am talking about `attr_reader :q` that is never used. Additionally, Is there some reason you need to override initialize? Can  it be completely inherited?

Comment (andjsmit) [inline app/authorities/sufia/authorities/local/find_works.rb]: @jcoyne Ok. The attr_reader can be removed. However, if I remove the `intitialize()`, the error above occurs.

Comment (andjsmit) [inline app/authorities/sufia/authorities/local/find_works.rb]: @jcoyne @jeremyf Tracking further back, the init is pulling in the subauthority.

Comment (jcoyne) [inline app/authorities/sufia/authorities/local/find_works.rb]: @andjsmit This is because you are attempting to https://github.com/projecthydra/sufia/pull/3015/files#diff-70cc36be61e57f96ec44887d4d62aa0fR120 register this as a subauthority. However it is not a subauthority. It is an authority in its own right.

Comment (jcoyne) [inline app/authorities/sufia/authorities/local/find_works.rb]: @andjsmit Apparently you are doing it the recommended way.   This still seems like a bug in QA.  I see that this has been worked around in the  past like this: https://github.com/curationexperts/alexandria/blob/7e6dfafefb2cd9bc17aa956208e8a86495e5b1eb/app/authorities/local_names.rb#L16-L17

Comment (jcoyne) [inline app/authorities/sufia/authorities/local/find_works.rb]: @andjsmit Okay, I looked more closely. Do not call `register_subauthority` and do not define an `initialize` method. This should just work because it is not a subauthority. It is equivalent to the built in tgnlang authority: https://github.com/projecthydra-labs/questioning_authority/blob/master/lib/qa/authorities/tgnlang.rb

File: app/forms/sufia/forms/work_form.rb [hunk 12]
```diff
@@ -49,6 +49,12 @@ def ordered_work_members
       model.ordered_members.to_a.select { |m| m.model_name.singular.to_sym != :file_set }
     end
 
+    # The in_work items
+    # @return [Array] All of the works in the in_works
```
Comment (jcoyne) [inline app/forms/sufia/forms/work_form.rb]: Can you rephrase "in the in_works" to "that this work is a member of"?

**Comment by mjgiarlo** [comment]
Re-review, @jeremyf?

File: app/forms/sufia/forms/work_form.rb (hunk 12 above)
Comment (andjsmit) [inline app/forms/sufia/forms/work_form.rb]: Description changed.

**Comment by mjgiarlo** [comment]
Well done, all!
